src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs

```diff
@@ -176,59 +166,32 @@ private TypeSymbol InferResultType(ImmutableArray<BoundSwitchExpressionArm> swit
             return commonType;
         }
 
-        private ImmutableArray<BoundSwitchExpressionArm> BindSwitchExpressionArms(SwitchExpressionSyntax node, Binder originalBinder, BindingDiagnosticBag diagnostics)
+        private ImmutableArray<BoundSwitchExpressionArm> BindSwitchExpressionArms(SwitchExpressionSyntax node, Binder originalBinder, BoundExpression inputExpression, BindingDiagnosticBag diagnostics)
         {
-            bool hasErrors = InputExpression.HasErrors;
             var builder = ArrayBuilder<BoundSwitchExpressionArm>.GetInstance();
+            (TypeSymbol inputType, uint valEscape) = GetInputTypeAndValEscape(inputExpression);
             foreach (var arm in node.Arms)
             {
-                var armBinder = originalBinder.GetBinder(arm);
-                var boundArm = armBinder.BindSwitchExpressionArm(arm, diagnostics);
+                var armBinder = originalBinder.GetRequiredBinder(arm);
+                Debug.Assert(inputExpression.Type is not null);
+                var boundArm = armBinder.BindSwitchExpressionArm(arm, inputType, valEscape, diagnostics);
                 builder.Add(boundArm);
             }
 
             return builder.ToImmutableAndFree();
         }
 
-        internal BoundExpression InputExpression
-        {
-            get
-            {
-                EnsureSwitchGoverningExpressionAndDiagnosticsBound();
-                Debug.Assert(_inputExpression != null);
-                return _inputExpression;
-            }
-        }
-
-        internal TypeSymbol SwitchGoverningType => InputExpression.Type;
-
-        internal uint SwitchGoverningValEscape => GetValEscape(InputExpression, LocalScopeDepth);
-
-        protected BindingDiagnosticBag InputExpressionDiagnostics
-        {
-            get
-            {
-                EnsureSwitchGoverningExpressionAndDiagnosticsBound();
-                Debug.Assert(_inputExpressionDiagnostics != null);
-                return _inputExpressionDiagnostics;
-            }
-        }
-
-        private void EnsureSwitchGoverningExpressionAndDiagnosticsBound()
+        internal (TypeSymbol GoverningType, uint GoverningValEscape) GetInputTypeAndValEscape(BoundExpression? inputExpression = null)
         {
-            if (_inputExpression == null)
-            {
-                var switchGoverningDiagnostics = new BindingDiagnosticBag();
-                var boundSwitchGoverningExpression = BindSwitchGoverningExpression(switchGoverningDiagnostics);
-                _inputExpressionDiagnostics = switchGoverningDiagnostics;
-                Interlocked.CompareExchange(ref _inputExpression, boundSwitchGoverningExpression, null);
```

Comment:
> The IOperation hook has revealed an additional bug here, that 07a0734 fixes. We were caching the governing expression for switch expressions in the switch expression binder, which caused the hook to assert for some of the newly added tests. I've removed the caching, and I'd suggest reviewing that single commit separately for ease.

I didn't understand how the caching of the bound governing expression resulted in a problem with the IOperation hook.

Reply:
> I didn't understand how the caching of the bound governing expression resulted in a problem with the IOperation hook. 

Because the node is cached, subsequent bindings will see the same node. It has already had its top-level nullability set, and the DebugVerifier asserts because it should only be seeing new nodes, not previously-analyzed nodes.

---
In reply to: [595616905](https://github.com/dotnet/roslyn/pull/51791#discussion_r595616905) [](ancestors = 595616905)